core/domain/stats_services.py

```diff
@@ -29,43 +29,66 @@
         exp_id, exp_version)
 
-    if event_type == feconf.EVENT_TYPE_START_EXPLORATION:
-        exploration_stats.num_starts_v2 += 1
-    elif event_type == feconf.EVENT_TYPE_ACTUAL_START_EXPLORATION:
-        exploration_stats.num_actual_starts_v2 += 1
-    elif event_type == feconf.EVENT_TYPE_COMPLETE_EXPLORATION:
-        exploration_stats.num_completions_v2 += 1
-    elif event_type == feconf.EVENT_TYPE_ANSWER_SUBMITTED:
-        exploration_stats.state_stats_mapping[
-            state_name].total_answers_count_v2 += 1
-        if update_params['feedback_is_useful']:
+    for event_param in event_params:
+        event_type = event_param['event_type']
+        update_params = event_param['update_params']
+        if event_type == feconf.EVENT_TYPE_START_EXPLORATION:
+            exploration_stats.num_starts_v2 += 1
+        elif event_type == feconf.EVENT_TYPE_ACTUAL_START_EXPLORATION:
+            exploration_stats.num_actual_starts_v2 += 1
+        elif event_type == feconf.EVENT_TYPE_COMPLETE_EXPLORATION:
+            exploration_stats.num_completions_v2 += 1
+        elif event_type == feconf.EVENT_TYPE_ANSWER_SUBMITTED:
+            state_name = event_param['state_name']
+            exploration_stats.state_stats_mapping[
+                state_name].total_answers_count_v2 += 1
+            if update_params['feedback_is_useful']:
+                exploration_stats.state_stats_mapping[
+                    state_name].useful_feedback_count_v2 += 1
+        elif event_type == feconf.EVENT_TYPE_STATE_HIT:
+            state_name = event_param['state_name']
+            exploration_stats.state_stats_mapping[
+                state_name].total_hit_count_v2 += 1
+            if update_params['is_first_hit']:
+                exploration_stats.state_stats_mapping[
+                    state_name].first_hit_count_v2 += 1
+        elif event_type == feconf.EVENT_TYPE_STATE_COMPLETED:
+            state_name = event_param['state_name']
             exploration_stats.state_stats_mapping[
-                state_name].useful_feedback_count_v2 += 1
-    elif event_type == feconf.EVENT_TYPE_STATE_HIT:
-        exploration_stats.state_stats_mapping[
-            state_name].total_hit_count_v2 += 1
-        if update_params['is_first_hit']:
+                state_name].num_completions_v2 += 1
+        elif event_type == feconf.EVENT_TYPE_SOLUTION_HIT:
+            state_name = event_param['state_name']
             exploration_stats.state_stats_mapping[
-                state_name].first_hit_count_v2 += 1
-    elif event_type == feconf.EVENT_TYPE_STATE_COMPLETED:
-        exploration_stats.state_stats_mapping[
-            state_name].num_completions_v2 += 1
-    elif event_type == feconf.EVENT_TYPE_SOLUTION_HIT:
-        exploration_stats.state_stats_mapping[
-            state_name].num_times_solution_viewed_v2 += 1
+                state_name].num_times_solution_viewed_v2 += 1
 
     save_stats_model(exploration_stats)
```

Comment:
Probably want to wrap this in a transaction for safety? (The taskqueue thing helps to ensure that this doesn't get overloaded, but you'd still want to guard against race conditions.)

Reply:
@seanlip I've wrapped it in a transaction. I tried to read up some stuff, but how do transactions help in this case? The whole concept seems a bit vague for me.

Reply:
The idea is to make the actions within the transaction atomic, so that they either all run at once or don't run at all. If operations collide then there will be a retry; see https://cloud.google.com/appengine/docs/standard/python/ndb/transactions. This helps address the issue of race conditions.